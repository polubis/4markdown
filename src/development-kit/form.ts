const isString = (value: unknown): value is string => typeof value === `string`;

const noEdgeSpaces = (value: string): boolean =>
  value.trim().length === value.length;

const nonNullable = <T>(value: T): value is NonNullable<T> =>
  value !== undefined && value !== null;

const notEmpty = (value: string): boolean => value.trim().length > 0;

const url = (value: string) =>
  /https?:\/\/(www\.)?[-a-zA-Z0-9@:%._+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_+.~#?&//=]*)/.test(
    value,
  );

const minLength =
  (limit: number) =>
  (value: string | any[]): boolean =>
    value.length >= limit;

const maxLength =
  (limit: number) =>
  (value: string | any[]): boolean =>
    value.length <= limit;

const chain =
  (...validators: ((value: any) => boolean)[]) =>
  (value: any): boolean => {
    for (const validator of validators) {
      if (!validator(value)) return false;
    }

    return true;
  };

type ValidatorsMap<Values extends Record<string, any>> = {
  [K in keyof Values]?: (value: Values[K]) => boolean;
};

type ValidationReport<Values extends Record<string, any>> = {
  ok: boolean;
  nok: boolean;
  result: Record<keyof Values, boolean>;
};

const report = <Values extends Record<string, any>>(
  values: Values,
  validators: ValidatorsMap<Values> = {},
): ValidationReport<Values> => {
  const result = {} as ValidationReport<Values>['result'];

  for (const key in values) {
    const validator = validators[key] ?? (() => true);
    result[key] = validator(values[key]);
  }

  for (const status in result) {
    if (!status) {
      return {
        result,
        ok: false,
        nok: true,
      };
    }
  }

  const nok = Object.values(result).some((status) => !status);

  return {
    result,
    ok: !nok,
    nok,
  };
};

const nickname = (value: string): boolean => /^[a-zA-Z0-9_-]+$/.test(value);

export type { ValidatorsMap, ValidationReport };
export {
  chain,
  isString,
  noEdgeSpaces,
  notEmpty,
  nonNullable,
  minLength,
  maxLength,
  nickname,
  report,
  url,
};
