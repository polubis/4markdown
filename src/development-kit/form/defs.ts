/* eslint-disable @typescript-eslint/no-explicit-any */
export type ValidatorResult<Identifier extends string> = Identifier | null;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ValuesBase = Record<string | number, any>;

export type ValidatorFn<Value, Result extends string> = (
  value: Value,
) => ValidatorResult<Result>;

export type ValidatorsSetup<Values extends ValuesBase> = {
  [Key in keyof Values]?: ValidatorFn<Values[Key], string>[];
};

export type ValuesKeys<Values extends ValuesBase> = (keyof Values)[];

export type ValidationResult<Values extends ValuesBase> = {
  [Key in keyof Values]: ValidatorResult<string>;
};

export interface FormState<Values extends ValuesBase> {
  values: Values;
  result: ValidationResult<Values>;
  touched: boolean;
  untouched: boolean;
  confirmed: boolean;
  unconfirmed: boolean;
  invalid: boolean;
  valid: boolean;
}

export type FormSubscriberAction =
  | `init`
  | `set`
  | `confirm`
  | `reset`
  | `reconfigure`;
export type FormSubscriber<Values extends ValuesBase> = (
  action: FormSubscriberAction,
  state: FormState<Values>,
) => void;
export type FormSubscription = () => void;

export interface Formable<Values extends ValuesBase> {
  state(): FormState<Values>;
  set(values: Partial<Values>): FormState<Values>;
  init(values: Values): FormState<Values>;
  confirm(): FormState<Values>;
  subscribe(subscriber: FormSubscriber<Values>): FormSubscription;
  reset(values?: Partial<Values>): FormState<Values>;
  reconfigure(
    values: Values,
    validators?: ValidatorsSetup<Values>,
  ): FormState<Values>;
}
