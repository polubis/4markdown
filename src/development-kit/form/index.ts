export * from './defs';
export * from './validators';
export * from './form';
export * from './core';
