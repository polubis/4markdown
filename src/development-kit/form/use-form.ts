import React from 'react';
import { ValidatorsSetup, form } from '.';

const useForm = <Values extends Record<string, any>>(
  values: Values,
  validators: ValidatorsSetup<Values> = {},
) => {
  const [instance] = React.useState(() => {
    const inst = form<Values>(validators);
    inst.init(values);
    return inst;
  });
  const [state, setState] = React.useState(instance.state);

  const set = React.useCallback(
    (values: Partial<Values>): void => {
      instance.set(values);
    },
    [instance],
  );

  const inject = React.useCallback(
    <Key extends keyof Values>(key: Key) => ({
      onChange: (
        e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
      ) => {
        instance.set({
          [key]: e.target.value,
        } as Partial<Values>);
      },
      name: key,
      value: state.values[key],
    }),
    [state, instance],
  );

  const confirm = React.useCallback((): void => {
    instance.confirm();
  }, [instance]);

  React.useEffect(() => {
    const unsubscribe = instance.subscribe((_, state) => setState(state));

    return () => {
      unsubscribe();
    };
  }, [instance]);

  return [
    state,
    { set, inject, confirm, subscribe: instance.subscribe },
  ] as const;
};

export { useForm };
