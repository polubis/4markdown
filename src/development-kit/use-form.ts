import React from 'react';
import { Formable, ValidatorsSetup, form } from './form';

const useForm = <Values extends Record<string, any>>(
  values: Values,
  validators: ValidatorsSetup<Values> = {},
) => {
  const [instance] = React.useState(() => {
    const inst = form<Values>(validators);
    inst.init(values);
    return inst;
  });
  const [state, setState] = React.useState(instance.state);

  const set: Formable<Values>['set'] = React.useCallback(
    (values) => instance.set(values),
    [instance],
  );

  const inject = React.useCallback(
    <Key extends keyof Values>(key: Key) => ({
      onChange: (
        e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
      ) => {
        instance.set({
          [key]: e.target.value,
        } as Partial<Values>);
      },
      name: key,
      value: state.values[key],
    }),
    [state, instance],
  );

  const confirm: Formable<Values>['confirm'] = React.useCallback(
    () => instance.confirm(),
    [instance],
  );

  const reconfigure: Formable<Values>['reconfigure'] = React.useCallback(
    (values, validators) => instance.reconfigure(values, validators),
    [instance],
  );

  const reset: Formable<Values>['reset'] = React.useCallback(
    (values) => instance.reset(values),
    [instance],
  );

  React.useEffect(() => {
    const unsubscribe = instance.subscribe((_, state) => setState(state));

    return () => {
      unsubscribe();
    };
  }, [instance]);

  return [
    state,
    { set, reconfigure, inject, reset, confirm, subscribe: instance.subscribe },
  ] as const;
};

export { useForm };
