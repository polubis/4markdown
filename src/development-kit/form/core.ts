import { ValidatorFn, ValidatorResult } from './defs';

export const chain =
  (...validators: ValidatorFn<any, string>[]) =>
  (value: any): ValidatorResult<string> => {
    for (const validator of validators) {
      const result = validator(value);
      if (result) return result;
    }

    return null;
  };

export const optional =
  (...validators: ((value: any) => ValidatorResult<string>)[]) =>
  (value: any): ValidatorResult<string> => {
    if (value === `` || value === undefined || value === null) return null;

    return chain(...validators)(value);
  };

export const validator =
  <ReturnLiteral extends string, ValueType>(
    literal: ReturnLiteral,
    predicate: (value: ValueType) => boolean,
  ) =>
  (value: ValueType): ValidatorResult<ReturnLiteral> =>
    predicate(value) ? null : literal;
