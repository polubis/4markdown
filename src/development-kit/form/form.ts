import { chain } from './core';
import type {
  ValidatorsSetup,
  ValuesBase,
  FormState,
  ValidationResult,
  FormSubscriber,
  FormSubscriberAction,
  Formable,
} from './defs';

export const form = <Values extends ValuesBase>(
  validators: ValidatorsSetup<Values> = {},
): Formable<Values> => {
  const subscriptions = new Map<string, FormSubscriber<Values>>();

  let state: FormState<Values>;
  let initialState: FormState<Values>;

  const validate = (
    values: Values,
  ): Pick<FormState<Values>, 'invalid' | 'valid' | 'result'> => {
    const result = {} as ValidationResult<Values>;
    let invalid = false;

    for (const key in values) {
      const value = values[key];
      const fns = validators[key] ?? [];
      result[key] = chain(...fns)(value);

      if (result[key] !== null) {
        invalid = true;
      }
    }

    return { result, invalid, valid: !invalid };
  };

  const setState = (newState: FormState<Values>): FormState<Values> => {
    state = {
      ...state,
      ...newState,
    };

    return state;
  };

  const confirm = (
    confirmed: boolean,
  ): Pick<FormState<Values>, 'confirmed' | 'unconfirmed'> => {
    return {
      confirmed,
      unconfirmed: !confirmed,
    };
  };

  const touch = (
    touched: boolean,
  ): Pick<FormState<Values>, 'touched' | 'untouched'> => {
    return {
      touched,
      untouched: !touched,
    };
  };

  const notify = (action: FormSubscriberAction): void => {
    const keys = subscriptions.keys();

    for (const key of keys) {
      const fn = subscriptions.get(key);
      fn?.(action, state);
    }
  };

  return {
    init: (values) => {
      const newState = setState({
        values,
        ...validate(values),
        ...confirm(false),
        ...touch(false),
      });
      notify(`init`);

      initialState = { ...newState };

      return newState;
    },
    set: (values) => {
      const newValues = {
        ...state.values,
        ...values,
      };

      const newState = setState({
        ...state,
        values: newValues,
        ...validate(newValues),
        ...touch(true),
      });
      notify(`set`);

      return newState;
    },
    confirm: () => {
      const newState = setState({
        ...state,
        ...validate(state.values),
        ...confirm(true),
      });
      notify(`confirm`);

      return newState;
    },
    reset: (values = {}) => {
      const newValues = {
        ...initialState.values,
        ...values,
      };

      const newState = setState({
        values: newValues,
        ...validate(newValues),
        ...confirm(false),
        ...touch(false),
      });

      notify(`reset`);

      return newState;
    },
    reconfigure: (values, newValidators = {}) => {
      validators = newValidators;

      const newState = setState({
        values,
        ...validate(values),
        ...confirm(false),
        ...touch(false),
      });

      notify(`reconfigure`);

      initialState = { ...newState };

      return newState;
    },
    subscribe: (subscriber) => {
      const key = new Date().toISOString();
      subscriptions.set(key, subscriber);

      return () => {
        subscriptions.delete(key);
      };
    },
    state: () => state,
  };
};
