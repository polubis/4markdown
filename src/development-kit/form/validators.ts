import { validator } from './core';

export const minLength = (limit: number) =>
  validator(`minLength`, (value: string | unknown[]) => value.length >= limit);

export const maxLength = (limit: number) =>
  validator(`maxLength`, (value: string | unknown[]) => value.length <= limit);

export const min = (limit: number) =>
  validator(`min`, (value: number) => value >= limit);

export const max = (limit: number) =>
  validator(`max`, (value: number) => value <= limit);

export const noEdgeSpaces = validator(
  `noEdgeSpaces`,
  (value: string) => value.trim().length === value.length,
);

export const notEmpty = validator(
  `notEmpty`,
  (value: string) => value.trim().length > 0,
);

export const url = validator(`url`, (value: string) =>
  /https?:\/\/(www\.)?[-a-zA-Z0-9@:%._+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_+.~#?&//=]*)/.test(
    value,
  ),
);

export const nickname = validator(`nickname`, (value: string) =>
  /^[a-zA-Z0-9_-]+$/.test(value),
);

export const base64 = validator(`base64`, (value: string) =>
  /^\s*data:([a-zA-Z]+\/[a-zA-Z]+)?(;base64)?,[a-zA-Z0-9+/]+={0,2}\s*$/.test(
    value,
  ),
);

export const isString = validator(
  `isString`,
  (value: unknown): value is string => typeof value === `string`,
);

export const name = validator(`name`, (value: string) =>
  /^[a-zA-Z0-9]+(?:\s[a-zA-Z0-9]+)*$/.test(value.trim()),
);

export const minWords = (count: number) =>
  validator(
    `minWords`,
    (value: string) => value.trim().split(` `).length >= count,
  );
