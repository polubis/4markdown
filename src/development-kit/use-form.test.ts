import { act, renderHook } from '@testing-library/react';
import { useForm } from './use-form';
import { maxLength, minLength, optional } from './form';
import { expect } from '@jest/globals';

describe(`Form management is integrated with React when: `, () => {
  interface SignInFormValues {
    username: string;
    password: string;
  }

  const createSignInFormValues = (
    username: SignInFormValues['username'],
    password: SignInFormValues['password'],
  ): SignInFormValues => ({
    username,
    password,
  });

  it(`allows to manage form`, () => {
    const { result } = renderHook(() =>
      useForm<SignInFormValues>(createSignInFormValues(``, ``), {
        username: [minLength(2), maxLength(5)],
        password: [optional(minLength(2), maxLength(5))],
      }),
    );

    expect(result.current[0]).toMatchSnapshot();

    act(() => {
      result.current[1].set({ username: `cos`, password: `s` });
    });

    expect(result.current[0]).toMatchSnapshot();

    act(() => {
      result.current[1].confirm();
    });

    expect(result.current[0]).toMatchSnapshot();

    act(() => {
      result.current[1].reset();
    });

    expect(result.current[0]).toMatchSnapshot();

    act(() => {
      result.current[1].set({ username: `cos`, password: `cos` });
    });

    expect(result.current[0]).toMatchSnapshot();

    act(() => {
      result.current[1].reconfigure({ username: ``, password: `` });
    });

    expect(result.current[0].values).toEqual(createSignInFormValues(``, ``));
    expect(result.current[0]).toMatchSnapshot();

    act(() => {
      result.current[1].set({ username: ``, password: `` });
    });

    expect(result.current[0]).toMatchSnapshot();
  });
});
