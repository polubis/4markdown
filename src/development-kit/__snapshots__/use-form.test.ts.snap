// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form management is integrated with React when:  allows to manage form 1`] = `
{
  "confirmed": false,
  "invalid": true,
  "result": {
    "password": null,
    "username": "minLength",
  },
  "touched": false,
  "unconfirmed": true,
  "untouched": true,
  "valid": false,
  "values": {
    "password": "",
    "username": "",
  },
}
`;

exports[`Form management is integrated with React when:  allows to manage form 2`] = `
{
  "confirmed": false,
  "invalid": true,
  "result": {
    "password": "minLength",
    "username": null,
  },
  "touched": true,
  "unconfirmed": true,
  "untouched": false,
  "valid": false,
  "values": {
    "password": "s",
    "username": "cos",
  },
}
`;

exports[`Form management is integrated with React when:  allows to manage form 3`] = `
{
  "confirmed": true,
  "invalid": true,
  "result": {
    "password": "minLength",
    "username": null,
  },
  "touched": true,
  "unconfirmed": false,
  "untouched": false,
  "valid": false,
  "values": {
    "password": "s",
    "username": "cos",
  },
}
`;

exports[`Form management is integrated with React when:  allows to manage form 4`] = `
{
  "confirmed": false,
  "invalid": true,
  "result": {
    "password": null,
    "username": "minLength",
  },
  "touched": false,
  "unconfirmed": true,
  "untouched": true,
  "valid": false,
  "values": {
    "password": "",
    "username": "",
  },
}
`;

exports[`Form management is integrated with React when:  allows to manage form 5`] = `
{
  "confirmed": false,
  "invalid": false,
  "result": {
    "password": null,
    "username": null,
  },
  "touched": true,
  "unconfirmed": true,
  "untouched": false,
  "valid": true,
  "values": {
    "password": "cos",
    "username": "cos",
  },
}
`;

exports[`Form management is integrated with React when:  allows to manage form 6`] = `
{
  "confirmed": false,
  "invalid": false,
  "result": {
    "password": null,
    "username": null,
  },
  "touched": false,
  "unconfirmed": true,
  "untouched": true,
  "valid": true,
  "values": {
    "password": "",
    "username": "",
  },
}
`;

exports[`Form management is integrated with React when:  allows to manage form 7`] = `
{
  "confirmed": false,
  "invalid": false,
  "result": {
    "password": null,
    "username": null,
  },
  "touched": true,
  "unconfirmed": true,
  "untouched": false,
  "valid": true,
  "values": {
    "password": "",
    "username": "",
  },
}
`;
